### PR TITLE
ID-1125 ECM Migrates Data from Bond

### PR DESCRIPTION
**Pull request by tlangs**
Jira: https://broadworkbench.atlassian.net/browse/ID-1125

We want to retire Bond and put its functionality into ECM. This presents a problem for users, since if they link an account in Bond, and then we cut over to ECM, they're linked account will go away! To resolve this, we can make ECM read from Bond's Datastore and dynamically update its own database if it finds any info. 

This is temporary logic that will exist for a maximum of 30 days, at which point all fence account links will have expired.

### CODE REVIEW

File: service/src/main/java/bio/terra/externalcreds/dataAccess/FenceAccountKeyDAO.java
Comment (samanehsan) [inline service/src/main/java/bio/terra/externalcreds/dataAccess/FenceAccountKeyDAO.java:37]: Does this also need a `@WithSpan` annotation? I realized I don't actually know what that does 🤔

File: service/src/main/java/bio/terra/externalcreds/models/BondFenceServiceAccountEntity.java
Comment (samanehsan) [inline service/src/main/java/bio/terra/externalcreds/models/BondFenceServiceAccountEntity.java:9]: This looks very similar to `FenceAccountKey` 🤔  How are they related?

File: service/src/main/java/bio/terra/externalcreds/services/FenceKeyRetriever.java
Comment (samanehsan) [inline service/src/main/java/bio/terra/externalcreds/services/FenceKeyRetriever.java]: I think this would be easier to read if this lock logic was extracted into a separate function 🔒

File: service/src/main/java/bio/terra/externalcreds/services/FenceKeyRetriever.java
Comment (samanehsan) [inline service/src/main/java/bio/terra/externalcreds/services/FenceKeyRetriever.java:146]: Maybe this should be a configurable value

File: service/src/test/java/bio/terra/externalcreds/services/ProviderTokenClientCacheTest.java
Comment (samanehsan) [inline service/src/test/java/bio/terra/externalcreds/services/ProviderTokenClientCacheTest.java:47]: Thanks for adding this test! 💯

File: service/src/main/java/bio/terra/externalcreds/services/FenceProviderService.java
Comment (samanehsan) [inline service/src/main/java/bio/terra/externalcreds/services/FenceProviderService.java:61]: When would the ECM linked account have an older expiration time than the one in Bond (since it would have been ported from the Bond db)?

File: service/src/test/java/bio/terra/externalcreds/BaseTest.java
Comment (samanehsan) [inline service/src/test/java/bio/terra/externalcreds/BaseTest.java:18]: What does this annotation do?

File: service/src/main/java/bio/terra/externalcreds/models/BondFenceServiceAccountEntity.java
Comment (samanehsan) [reply service/src/main/java/bio/terra/externalcreds/models/BondFenceServiceAccountEntity.java:9]: Oh it looks like `BondFenceServiceAccountEntity` is the model for the info we get from Bond and `FenceAccountKey` is just used for the ECM representation?

File: service/src/main/java/bio/terra/externalcreds/services/FenceProviderService.java
Comment (tlangs) [reply service/src/main/java/bio/terra/externalcreds/services/FenceProviderService.java:61]: If, somehow, someone connected to Bond directly to renew their link. Probably won't happen, but still a possibility.

File: service/src/main/java/bio/terra/externalcreds/models/BondFenceServiceAccountEntity.java
Comment (tlangs) [reply service/src/main/java/bio/terra/externalcreds/models/BondFenceServiceAccountEntity.java:9]: Exactly! The `BondDatastoreDAO` is bond-specific, and the `BondService` handles converting Bond models to ECM ones. I didn't want to encode a lot of business logic in the DAO layer.

File: service/src/main/java/bio/terra/externalcreds/dataAccess/FenceAccountKeyDAO.java
Comment (tlangs) [reply service/src/main/java/bio/terra/externalcreds/dataAccess/FenceAccountKeyDAO.java:37]: I believe it does

Comment (tlangs) [reply service/src/main/java/bio/terra/externalcreds/dataAccess/FenceAccountKeyDAO.java:37]: `@WithSpan` allows us to trace the performance of the DB query separate from the API request that triggered it.

File: service/src/test/java/bio/terra/externalcreds/BaseTest.java
Comment (tlangs) [reply service/src/test/java/bio/terra/externalcreds/BaseTest.java:18]: It makes sure that Spring throws out its state after each test file. Without this, Spring keeps around its db connections for completed tests, and we end up running out of db connections!

File: service/src/main/java/bio/terra/externalcreds/services/FenceKeyRetriever.java
Comment (tlangs) [reply service/src/main/java/bio/terra/externalcreds/services/FenceKeyRetriever.java:146]: Probably. I'll have to wait until the provider config is merged to do that though.

File: service/src/main/java/bio/terra/externalcreds/services/FenceKeyRetriever.java
Comment (samanehsan) [inline service/src/main/java/bio/terra/externalcreds/services/FenceKeyRetriever.java:67]: If the fence account retrieval/upsert fails it won't clean up the lock. What about deleting the existing lock if it is expired? Or updating `obtainLockAndGetNewKey` so that it deletes the lock even if it fails?

Comment (samanehsan) [reply service/src/main/java/bio/terra/externalcreds/services/FenceKeyRetriever.java:67]: Also thanks for splitting up this function! I find it much easier to follow what it's doing!

File: service/src/main/java/bio/terra/externalcreds/services/FenceProviderService.java
Comment (samanehsan) [reply service/src/main/java/bio/terra/externalcreds/services/FenceProviderService.java:61]: Oh I see!

File: service/src/main/java/bio/terra/externalcreds/services/FenceKeyRetriever.java
Comment (tlangs) [reply service/src/main/java/bio/terra/externalcreds/services/FenceKeyRetriever.java:67]: Just made that happen!

File: service/src/main/java/bio/terra/externalcreds/controllers/OauthApiController.java
Comment (Ghost-in-a-Jar) [inline service/src/main/java/bio/terra/externalcreds/controllers/OauthApiController.java:47]: yay for pattern matching

File: service/src/main/java/bio/terra/externalcreds/dataAccess/BondDatastoreDAO.java
Comment (Ghost-in-a-Jar) [inline service/src/main/java/bio/terra/externalcreds/dataAccess/BondDatastoreDAO.java:63]: does this need to be decoded or anything

Comment (Ghost-in-a-Jar) [reply service/src/main/java/bio/terra/externalcreds/dataAccess/BondDatastoreDAO.java:63]: or the key i mean

**Review by Ghost-in-a-Jar — APPROVED** [review]
This looks good i think you did a good job of covering test cases. One thing that would be helpful for reviewing such a large pr in the future is to add some more detail about the logic flow to the description and maybe inline pr comments.

**Review by samanehsan — APPROVED** [review]
I think this looks good once it's up-to-date with the changes I just merged! :octocat:

**Comment by tlangs** [comment]
Yesss I got coverage!

**Pull request merged**
